src/models/models.cpp

```diff
@@ -316,6 +321,53 @@ std::vector<int> Model::set_input(std::vector<std::vector<int32_t>> &inputIds_,
     return seqIDs;
 }
 
+std::vector<int> Model::set_input(std::vector<std::vector<int32_t>> &inputIds_, std::vector<int> seqIDs, int maxLen) {
+    Messenger &messenger = Messenger::getInstance();
+    SequencePool &seqPool = SequencePool::getInstance();
+    KVCacheMgr &kvCacheMgr = KVCacheMgr::instance();
+    workingGroup.clear();
+    batchSize = inputIds_.size();
+
+    maxLen = std::min(maxLen, this->maxSeqLen);
+
+    if (messenger.getSize() > 1) {
+        // TODO: Sync input and sampling param in distributed mode.
+        // [batch_size, total_length, seqID_size, maxLen]
+    }
+    if (seqIDs.empty()) {
+        // Prompt(1st token)
+        // Create seq meta for inputs and return seq IDs
+        for (int i = 0; i < batchSize; i++) {
+            auto group = seqPool.newGroupMeta(inputIds_[i]);
+            workingGroup.push_back(group);
+            seqIDs.push_back(group->getGroupID());
+            kvCacheMgr.addSequence(group->getGroupID(), maxLen);
+        }
+    } else {
+        // Decode(next token)
+        // Update seq meta with inputs and return seq IDs
+        if (inputIds_.size() != seqIDs.size()) {
+            printf("[ERROR] Input size and seqIDs size mismatch.\n");
+            exit(-1);
+        }
+        for (int i = 0; i < batchSize; i++) {
+            auto group = seqPool.get(seqIDs[i]);
+            if (group == nullptr) {
+                // TODO: Address beam search case.
+                printf("[ERROR] Sequence ID %d not found.\n", seqIDs[i]);
+                exit(-1);
+            }
+            group->get(0)->stepForward(inputIds_[i][0]);
```

Comment:
Why not put 'stepForward' to forward serial function?

Reply:
since generate() already called stepForward, it may cause duplicate 2 times calling.
从语义上说，set_input不应该调stepForward, stepForward可能会将相关结构被修改为forward之后的状态，这样forward的时候就不对了

Reply:
我写了个新的函数来用`std::vector<int> set_input(std::vector<std::vector<int32_t>> &inputIds_, std::vector<int> seqIDs, SearcherConfig &config_, const std::vector<std::vector<int>> &stopWordsList_ = {})`， 这个不会调用`stepforward`。`    std::vector<int> set_input(std::vector<std::vector<int32_t>> &inputIds_, std::vector<int> seqIDs = {}, int maxLen = -1);`还是保留，和注释保持一致，只有在`model.forward()`的时候使用，这是给vllm使用的。如果只调用`model.forward()`，xft不会处理search，所以在那次调用的时候不会知道next token是什么。